src/DistributedComputations/distributed_fft_based_poisson_solver.jl

```diff
@@ -55,155 +47,159 @@ Algorithm for two-dimensional decompositions
-    x̂ = b̂ = last(solver.storage)
     λ = solver.eigenvalues
-    @. x̂ = - b̂ / (λ[1] + λ[2] + λ[3])
+    x̂ = b̂ = parent(storage.xfield)
+    
+    launch!(arch, storage.xfield.grid, :xyz, _solve_poisson_in_spectral_space!, x̂, b̂, λ[1], λ[2], λ[3])
 
     # Set the zeroth wavenumber and volume mean, which are undetermined
     # in the Poisson equation, to zero.
-    if MPI.Comm_rank(multi_arch.communicator) == 0
-        # This is an assumption: we *hope* PencilArrays allocates in this way
-        parent(x̂)[1, 1, 1] = 0
+    if arch.local_rank == 0
+        @allowscalar x̂[1, 1, 1] = 0
     end
 
-    # Apply backward transforms to x̂ = last(solver.storage).
-    solver.plan \ solver.storage
-
-    # xc is the backward transform of x̂.
-    xc = first(solver.storage)
+    # Apply backward transforms to x̂ = parent(storage.xfield).
+    solver.plan.backward.x!(parent(storage.xfield), buffer.x)
+    transpose_x_to_y!(storage) # copy data from storage.xfield to storage.yfield
+    solver.plan.backward.y!(parent(storage.yfield), buffer.y)
+    transpose_y_to_z!(storage) # copy data from storage.yfield to storage.zfield
+    solver.plan.backward.z!(parent(storage.zfield), buffer.z) # last backwards transform is in z
 
     # Copy the real component of xc to x.
     launch!(arch, solver.local_grid, :xyz,
-            copy_permuted_real_component!, x, parent(xc), solver.input_permutation)
+            _copy_real_component!, x, parent(storage.zfield))
 
     return x
 end
 
-const ZXYPermutation = Permutation{(3, 1, 2), 3}
-const ZYXPermutation = Permutation{(3, 2, 1), 3}
-
-@kernel function copy_permuted_real_component!(ϕ, ϕc, ::ZXYPermutation)
+@kernel function _solve_poisson_in_spectral_space!(x̂, b̂, λx, λy, λz)
     i, j, k = @index(Global, NTuple)
-    @inbounds ϕ[i, j, k] = real(ϕc[k, i, j])
+    @inbounds x̂[i, j, k] = - b̂[i, j, k] / (λx[i] + λy[j] + λz[k])
 end
 
-@kernel function copy_permuted_real_component!(ϕ, ϕc, ::ZYXPermutation)
+@kernel function _copy_real_component!(ϕ, ϕc)
     i, j, k = @index(Global, NTuple)
-    @inbounds ϕ[i, j, k] = real(ϕc[k, j, i])
+    @inbounds ϕ[i, j, k] = real(ϕc[i, j, k])
+end
+
+# TODO: bring up to speed the PCG to remove this error
+validate_global_grid(global_grid) = 
+        throw(ArgumentError("Grids other than the RectilinearGrid are not supported in the Distributed NonhydrostaticModels"))
+
+function validate_global_grid(global_grid::RectilinearGrid) 
+    TX, TY, TZ = topology(global_grid)
+
+    if (TY == Bounded && TZ == Periodic) || (TX == Bounded && TY == Periodic) || (TX == Bounded && TZ == Periodic)
+        throw(ArgumentError("NonhydrostaticModels on Distributed grids do not support topology ($TX, $TY, $TZ) at the moment.
+                             TZ Periodic requires also TY and TX to be Periodic,
+                             while TY Periodic requires also TX to be Periodic.
+                             Please rotate the domain to obtain the required topology"))
```

Comment:
Are these requirements in any docstring that's easy to find?

Reply:
yeah they are in the restrictions section